tests/storage/test_base.py

```diff
@@ -396,7 +397,28 @@ def test_upsert(self) -> Generator["defer.Deferred[object]", object, None]:
         self.assertTrue(result)
 
     @defer.inlineCallbacks
-    def test_upsert_with_insert(
+    def test_upsert_no_values(
```

Comment:
Are you sure this is testing the "emulation" logic? I thought that a table-level lock was required: https://github.com/matrix-org/synapse/blob/cfb6d38c47711b8dfaf0125353aec88d16708b97/synapse/storage/database.py#L1188-L1189

(maybe this is PG only since there is only ever a single sqlite writer?)

Reply:
Ahhh: https://github.com/matrix-org/synapse/blob/cfb6d38c47711b8dfaf0125353aec88d16708b97/synapse/storage/database.py#L1220-L1221

Reply:
https://github.com/matrix-org/synapse/blob/cfb6d38c47711b8dfaf0125353aec88d16708b97/synapse/storage/database.py#L1271-L1288 I think you may need to patch/mock `_unsafe_to_upsert_tables`?

Reply:
It does -- it adds the table into the `_unsafe_to_upsert_tables` set.

Reply:
> Are you sure this (and others) is testing the "emulation" logic? I thought that a table-level lock was required:

That gets called if `lock` is set to `True`, which is the default:

https://github.com/matrix-org/synapse/blob/80a7f75e4dc4d8612c61f02faa8b17bcc8af54d0/synapse/storage/database.py#L1315-L1317

tl;dr -- it uses a different API on the engine. For [sqlite this is a no-op](https://github.com/matrix-org/synapse/blob/4f1840a88ad3a93244fc23149c56245704eab824/synapse/storage/engines/sqlite.py#L118-L119), but for [postgres it does something](https://github.com/matrix-org/synapse/blob/4f1840a88ad3a93244fc23149c56245704eab824/synapse/storage/engines/postgres.py#L197-L198).

I think this is because [`assert_called_with`](https://docs.python.org/3/library/unittest.mock.html#unittest.mock.Mock.assert_called_with) checks the *last* call only, ignoring prior calls.

Reply:
I switched everything to `assert_called_once_with`, then fixed the Postgres-specific ones to add the lock call.